We should handle device disconnection and device name change (the case of bad-contact)

Useful log section from a disconnection:

```bash
2023-04-28T14:16:33.647333Z DEBUG actix-rt|system:0|arbiter:0 ThreadId(37) HTTP request{http.method=POST http.route=/streams http.flavor=1.1 http.scheme=http http.host=0.0.0.0:6020 http.client_ip=127.0.0.1 http.user_agent=Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/112.0 http.target=/streams otel.name=HTTP POST /streams otel.kind="server" request_id=32d9c728-0329-4d01-830b-b6e086d9a9f5}:add_stream_and_start{video_and_stream_information=VideoAndStreamInformation { name: "/dev/video2 - H264 USB Camera: USB Camera", stream_information: StreamInformation { endpoints: [Url { scheme: "udp", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("0.0.0.0")), port: Some(5600), path: "", query: None, fragment: None }], configuration: Video(VideoCaptureConfiguration { encode: H264, height: 1080, width: 1920, frame_interval: FrameInterval { numerator: 1, denominator: 30 } }), extended_configuration: Some(ExtendedConfiguration { thermal: false }) }, video_source: Local(VideoSourceLocal { name: "H264 USB Camera: USB Camera", device_path: "/dev/video2", typ: Usb("usb-0000:09:00.3-3.2") }) }}:add_stream{stream=Stream { id: 77bb288d-6485-4b0b-84cb-2d78fd6774c4, pipeline: V4l(V4lPipeline { state: PipelineState { pipeline_id: 77bb288d-6485-4b0b-84cb-2d78fd6774c4, pipeline: Pipeline { inner: TypedObjectRef { inner: 0x7fb9c0089620, type: GstPipeline } }, sink_tee: Element { inner: TypedObjectRef { inner: 0x7fb9c00897e0, type: GstTee } }, sinks: {022346b5-5130-4abe-8a27-8f3df1b8c89a: Image(ImageSink { sink_id: 022346b5-5130-4abe-8a27-8f3df1b8c89a, pipeline: Pipeline { inner: TypedObjectRef { inner: 0x7fb9c019c7b0, type: GstPipeline } }, queue: Element { inner: TypedObjectRef { inner: 0x7fb9c009f020, type: GstQueue } }, proxysink: Element { inner: TypedObjectRef { inner: 0x7fb9c00e1d10, type: GstProxySink } }, _proxysrc: Element { inner: TypedObjectRef { inner: 0x7fb9c00e2410, type: GstProxySrc } }, _transcoding_elements: [Element { inner: TypedObjectRef { inner: 0x7fb9c00e7bd0, type: GstRtpH264Depay } }, Element { inner: TypedObjectRef { inner: 0x7fb9c00ead40, type: GstH264Parse } }, Element { inner: TypedObjectRef { inner: 0x7fb9c00efb80, type: GstIdentity } }, Element { inner: TypedObjectRef { inner: 0x7fb9c018e820, type: avdec_h264 } }, Element { inner: TypedObjectRef { inner: 0x7fb9c0199890, type: GstVideoConvert } }], appsink: AppSink { inner: TypedObjectRef { inner: 0x7fb9c019bdb0, type: GstAppSink } }, tee_src_pad: Some(Pad { inner: TypedObjectRef { inner: 0x7fb9c019ea70, type: GstTeePad } }), flat_samples_sender: broadcast::Sender, pad_blocker: Mutex { data: None, poisoned: false, .. }, _pipeline_runner: PipelineRunner { pipeline_weak: WeakRef(GWeakRef @ 0x7fb9c01a0c50 { priv_: GWeakRef_priv @ 0x7fb9c01a0c50 { p: 0x7fb9c019c7b0 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>), start_signal_sender: broadcast::Sender, killswitch_sender: broadcast::Sender, _killswitch_receiver: broadcast::Receiver, _watcher_thread_handle: JoinHandle { .. } }, thumbnails: Mutex { data: CachedThumbnails { map: {} }, poisoned: false, .. } }), 51c01fb8-306d-4145-9001-dc39ad25f782: Udp(UdpSink { sink_id: 51c01fb8-306d-4145-9001-dc39ad25f782, pipeline: Pipeline { inner: TypedObjectRef { inner: 0x7fb9c009a600, type: GstPipeline } }, queue: Element { inner: TypedObjectRef { inner: 0x7fb9c008ba20, type: GstQueue } }, proxysink: Element { inner: TypedObjectRef { inner: 0x7fb9c008ef80, type: GstProxySink } }, _proxysrc: Element { inner: TypedObjectRef { inner: 0x7fb9c008fc70, type: GstProxySrc } }, _udpsink: Element { inner: TypedObjectRef { inner: 0x7fb9c00980e0, type: GstMultiUDPSink } }, udpsink_sink_pad: Pad { inner: TypedObjectRef { inner: 0x7fb9c00984f0, type: GstPad } }, tee_src_pad: Some(Pad { inner: TypedObjectRef { inner: 0x7fb9c009f660, type: GstTeePad } }), addresses: [Url { scheme: "udp", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("0.0.0.0")), port: Some(5600), path: "", query: None, fragment: None }], _pipeline_runner: PipelineRunner { pipeline_weak: WeakRef(GWeakRef @ 0x7fb9c00943c0 { priv_: GWeakRef_priv @ 0x7fb9c00943c0 { p: 0x7fb9c009a600 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>), start_signal_sender: broadcast::Sender, killswitch_sender: broadcast::Sender, _killswitch_receiver: broadcast::Receiver, _watcher_thread_handle: JoinHandle { .. } } })}, pipeline_runner: PipelineRunner { pipeline_weak: WeakRef(GWeakRef @ 0x7fb9c0059f70 { priv_: GWeakRef_priv @ 0x7fb9c0059f70 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>), start_signal_sender: broadcast::Sender, killswitch_sender: broadcast::Sender, _killswitch_receiver: broadcast::Receiver, _watcher_thread_handle: JoinHandle { .. } }, _watcher_thread_handle: JoinHandle { id: Id(60) } } }), video_and_stream_information: VideoAndStreamInformation { name: "/dev/video2 - H264 USB Camera: USB Camera", stream_information: StreamInformation { endpoints: [Url { scheme: "udp", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("0.0.0.0")), port: Some(5600), path: "", query: None, fragment: None }], configuration: Video(VideoCaptureConfiguration { encode: H264, height: 1080, width: 1920, frame_interval: FrameInterval { numerator: 1, denominator: 30 } }), extended_configuration: Some(ExtendedConfiguration { thermal: false }) }, video_source: Local(VideoSourceLocal { name: "H264 USB Camera: USB Camera", device_path: "/dev/video2", typ: Usb("usb-0000:09:00.3-3.2") }) }, mavlink_camera: None }}:update_settings: src/settings/manager.rs:130: Settings saved: SettingsStruct {
    header: HeaderSettingsFile {
        name: "Camera Manager",
        version: 0,
    },
    mavlink_endpoint: None,
    streams: [
        VideoAndStreamInformation {
            name: "/dev/video2 - H264 USB Camera: USB Camera",
            stream_information: StreamInformation {
                endpoints: [
                    Url {
                        scheme: "udp",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "0.0.0.0",
                            ),
                        ),
                        port: Some(
                            5600,
                        ),
                        path: "",
                        query: None,
                        fragment: None,
                    },
                ],
                configuration: Video(
                    VideoCaptureConfiguration {
                        encode: H264,
                        height: 1080,
                        width: 1920,
                        frame_interval: FrameInterval {
                            numerator: 1,
                            denominator: 30,
                        },
                    },
                ),
                extended_configuration: Some(
                    ExtendedConfiguration {
                        thermal: false,
                    },
                ),
            },
            video_source: Local(
                VideoSourceLocal {
                    name: "H264 USB Camera: USB Camera",
                    device_path: "/dev/video2",
                    typ: Usb(
                        "usb-0000:09:00.3-3.2",
                    ),
                },
            ),
        },
    ],
}
2023-04-28T14:16:33.647777Z  INFO actix-rt|system:0|arbiter:0 ThreadId(37) /home/joaoantoniocardoso/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.3.1/src/middleware/logger.rs:421: 127.0.0.1 "POST /streams HTTP/1.1" 200 818 "http://0.0.0.0:6020/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/112.0" 0.052198    
2023-04-28T14:16:33.678250Z  INFO actix-rt|system:0|arbiter:0 ThreadId(37) /home/joaoantoniocardoso/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.3.1/src/middleware/logger.rs:421: 127.0.0.1 "GET /v4l HTTP/1.1" 200 14215 "http://0.0.0.0:6020/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/112.0" 0.027524    
2023-04-28T14:16:33.681434Z  INFO actix-rt|system:0|arbiter:0 ThreadId(37) /home/joaoantoniocardoso/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-4.3.1/src/middleware/logger.rs:421: 127.0.0.1 "GET /streams HTTP/1.1" 200 818 "http://0.0.0.0:6020/" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/112.0" 0.000373    
2023-04-28T14:16:33.706315Z DEBUG PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:120: Starting signal received in Pipeline 77bb288d-6485-4b0b-84cb-2d78fd6774c4
2023-04-28T14:16:34.865443Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 1
2023-04-28T14:16:34.965747Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 2
2023-04-28T14:16:35.066004Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 3
2023-04-28T14:16:35.166284Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 4
2023-04-28T14:16:35.266591Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 5
2023-04-28T14:16:35.366816Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 6
2023-04-28T14:16:35.467135Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 7
2023-04-28T14:16:35.567454Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 8
2023-04-28T14:16:35.667732Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 9
2023-04-28T14:16:35.768010Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 10
2023-04-28T14:17:13.602886Z ERROR PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:185: Error from Some("/GstPipeline:pipeline0/GstV4l2Src:v4l2src0"): Could not read from resource. (Some("../gstreamer/subprojects/gst-plugins-good/sys/v4l2/gstv4l2bufferpool.c(1198): gst_v4l2_buffer_pool_poll (): /GstPipeline:pipeline0/GstV4l2Src:v4l2src0:\npoll error 1: Resource temporarily unavailable (11)"))
2023-04-28T14:17:13.703201Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 1
2023-04-28T14:17:13.703313Z ERROR PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:185: Error from Some("/GstPipeline:pipeline0/GstV4l2Src:v4l2src0"): Failed to allocate a buffer (Some("../gstreamer/subprojects/gst-plugins-good/sys/v4l2/gstv4l2src.c(1374): gst_v4l2src_create (): /GstPipeline:pipeline0/GstV4l2Src:v4l2src0"))
2023-04-28T14:17:13.803489Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 2
2023-04-28T14:17:13.803642Z ERROR PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:185: Error from Some("/GstPipeline:pipeline0/GstV4l2Src:v4l2src0"): Internal data stream error. (Some("../gstreamer/subprojects/gstreamer/libs/gst/base/gstbasesrc.c(3132): gst_base_src_loop (): /GstPipeline:pipeline0/GstV4l2Src:v4l2src0:\nstreaming stopped, reason error (-5)"))
2023-04-28T14:17:13.903854Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:154: Position did not change 3
2023-04-28T14:17:13.903963Z  WARN PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7fb9c0077800 { priv_: GWeakRef_priv @ 0x7fb9c0077800 { p: 0x7fb9c0089620 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4 killswitch_receiver=broadcast::Receiver start_signal_receiver=broadcast::Receiver allow_block=false}: src/stream/pipeline/runner.rs:181: Received EndOfStream: Eos(
    Message {
        ptr: 0x00007fb98400d2e0,
        type: "eos",
        seqnum: 58,
        src: Some(
            "pipeline0",
        ),
        structure: None,
    },
)
2023-04-28T14:17:13.904056Z  INFO PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) src/stream/pipeline/runner.rs:51: PipelineWatcher ended with no error.
2023-04-28T14:17:13.904096Z  INFO PipelineRunner-77bb288d-6485-4b0b-84cb-2d78fd6774c4 ThreadId(46) src/stream/pipeline/runner.rs:58: Error sent to killswitch channel!
2023-04-28T14:17:13.904180Z DEBUG                         actix-rt|system:0|arbiter:0 ThreadId(37) src/stream/pipeline/mod.rs:142: Killswitch received as 77bb288d-6485-4b0b-84cb-2d78fd6774c4 from PipelineState's watcher. Reason: "Normal ending"
2023-04-28T14:17:13.904415Z DEBUG                         actix-rt|system:0|arbiter:0 ThreadId(37) remove_stream{stream_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4}:update_settings: src/settings/manager.rs:130: Settings saved: SettingsStruct {
    header: HeaderSettingsFile {
        name: "Camera Manager",
        version: 0,
    },
    mavlink_endpoint: None,
    streams: [],
}
2023-04-28T14:17:13.908235Z  INFO                         actix-rt|system:0|arbiter:0 ThreadId(37) remove_stream{stream_id=77bb288d-6485-4b0b-84cb-2d78fd6774c4}: src/stream/manager.rs:547: Stream 77bb288d-6485-4b0b-84cb-2d78fd6774c4 successfully removed!
2023-04-28T14:17:13.908518Z  INFO                         actix-rt|system:0|arbiter:0 ThreadId(37) src/stream/pipeline/mod.rs:146: Pipeline 77bb288d-6485-4b0b-84cb-2d78fd6774c4 removed. Reason: Normal ending
2023-04-28T14:17:13.935171Z ERROR PipelineRunner-51c01fb8-306d-4145-9001-dc39ad25f782 ThreadId(47) src/stream/pipeline/runner.rs:48: PipelineWatcher ended with error: Failed receiving start signal: Closed
2023-04-28T14:17:13.935273Z ERROR PipelineRunner-51c01fb8-306d-4145-9001-dc39ad25f782 ThreadId(47) src/stream/pipeline/runner.rs:56: Failed to broadcast error from PipelineWatcher. Reason: channel closed
2023-04-28T14:17:13.973084Z ERROR PipelineRunner-022346b5-5130-4abe-8a27-8f3df1b8c89a ThreadId(48) src/stream/pipeline/runner.rs:48: PipelineWatcher ended with error: Failed receiving start signal: Closed
2023-04-28T14:17:13.973150Z ERROR PipelineRunner-022346b5-5130-4abe-8a27-8f3df1b8c89a ThreadId(48) src/stream/pipeline/runner.rs:56: Failed to broadcast error from PipelineWatcher. Reason: channel closed
```


Useful log section of a bad-contact:

--- a/src/stream/manager.rs
+++ b/src/stream/manager.rs
@@ -53,7 +53,13 @@ impl Manager {
         let video_and_stream_informations = self
             .streams
             .values()
-            .map(|stream| stream.video_and_stream_information.clone())
+            .filter_map(|stream| match stream.state.lock() {
+                Ok(guard) => Some(guard.video_and_stream_information.clone()),
+                Err(error) => {
+                    error!("Failed locking a Mutex. Reason: {error}");
+                    None
+                }
+            })
             .collect::<Vec<VideoAndStreamInformation>>();
 
         settings::manager::set_streams(video_and_stream_informations.as_slice());
@@ -126,7 +132,7 @@ pub fn start_default() -> Result<()> {
 
     // Update all local video sources to make sure that they are available
     let mut candidates = video_source::cameras_available();
-    update_devices(&mut streams, &mut candidates);
+    update_devices(&mut streams, &mut candidates, true);
 
     // Remove all invalid video_sources
     let streams: Vec<VideoAndStreamInformation> = streams
@@ -146,9 +152,10 @@ pub fn start_default() -> Result<()> {
 }
 
 #[instrument(level = "debug")]
-fn update_devices(
+pub fn update_devices(
     streams: &mut Vec<VideoAndStreamInformation>,
     candidates: &mut Vec<VideoSourceType>,
+    verbose: bool,
 ) {
     for stream in streams {
         let VideoSourceType::Local(source) = &mut stream.video_source else {
@@ -179,7 +186,9 @@ fn update_devices(
             Err(reason) => {
                 // Invalidate the device
                 source.device_path = "".into();
-                warn!("Device {source:?} was invalidated. Reason: {reason:?}");
+                if verbose {
+                    warn!("Device {source:?} was invalidated. Reason: {reason:?}");
+                };
             }
             _ => (),
         }
@@ -203,9 +212,16 @@ pub fn get_first_sdp_from_source(source: String) -> ClonableResult<gst_sdp::SDPM
         .streams
         .values()
         .find_map(|stream| {
-            if stream.video_and_stream_information.video_source.inner().source_string() == source {
-                stream
-                    .pipeline
+            let state = match stream.state.lock() {
+                Ok(guard) => guard,
+                Err(error) => {
+                    error!("Failed locking a Mutex. Reason: {error}");
+                    return None;
+                },
+            };
+
+            if state.video_and_stream_information.video_source.inner().source_string() == source {
+                state.pipeline
                     .inner_state_as_ref()
                     .sinks
                     .values()
@@ -266,8 +282,15 @@ pub async fn get_jpeg_thumbnail_from_source(
                         }
                     };
                     let Some(stream) = manager.streams.values().find(|stream| {
-                        stream
-                            .video_and_stream_information
+                        let state = match stream.state.lock() {
+                            Ok(guard) => guard,
+                            Err(error) => {
+                                error!("Failed locking a Mutex. Reason: {error}");
+                                return false;
+                            },
+                        };
+
+                        state.video_and_stream_information
                             .video_source
                             .inner()
                             .source_string()
@@ -276,7 +299,15 @@ pub async fn get_jpeg_thumbnail_from_source(
                         return None
                     };
 
-                    let mut sinks = futures::stream::iter(stream.pipeline.inner_state_as_ref().sinks.values());
+                    let state = match stream.state.lock() {
+                        Ok(guard) => guard,
+                        Err(error) => {
+                            error!("Failed locking a Mutex. Reason: {error}");
+                            return None;
+                        },
+                    };
+
+                    let mut sinks = futures::stream::iter(state.pipeline.inner_state_as_ref().sinks.values());
                     let Some(Sink::Image(image_sink)) = sinks.find(|sink| matches!(sink, Sink::Image(_))).await else {
                         return None;
                     };
@@ -304,7 +335,14 @@ pub fn add_stream_and_start(video_and_stream_information: VideoAndStreamInformat
         Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
     };
     for stream in manager.streams.values() {
-        stream
+        let state = match stream.state.lock() {
+            Ok(guard) => guard,
+            Err(error) => {
+                return Err(anyhow!("Failed locking a Mutex. Reason: {error}"));
+            }
+        };
+
+        state
             .video_and_stream_information
             .conflicts_with(&video_and_stream_information)?;
     }
@@ -323,7 +361,15 @@ pub fn remove_stream_by_name(stream_name: &str) -> Result<()> {
         Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
     };
     if let Some(stream_id) = &manager.streams.iter().find_map(|(id, stream)| {
-        if stream.video_and_stream_information.name == *stream_name {
+        let state = match stream.state.lock() {
+            Ok(guard) => guard,
+            Err(error) => {
+                error!("Failed locking a Mutex. Reason: {error}");
+                return None;
+            }
+        };
+
+        if state.video_and_stream_information.name == *stream_name {
             return Some(*id);
         }
         None
@@ -362,7 +408,7 @@ impl WebRTCSessionManagementInterface for Manager {
         };
 
         let sink = Sink::WebRTC(WebRTCSink::try_new(bind, sender)?);
-        stream.pipeline.add_sink(sink)?;
+        stream.state.lock().unwrap().pipeline.add_sink(sink)?;
         debug!("WebRTC session created: {session_id:?}");
 
         Ok(session_id)
@@ -383,7 +429,12 @@ impl WebRTCSessionManagementInterface for Manager {
             .get_mut(&bind.producer_id)
             .context(format!("Producer {:?} not found", bind.producer_id))?;
 
-        stream
+        let mut state = match stream.state.lock() {
+            Ok(guard) => guard,
+            Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
+        };
+
+        state
             .pipeline
             .inner_state_mut()
             .remove_sink(&bind.session_id)
@@ -404,10 +455,18 @@ impl WebRTCSessionManagementInterface for Manager {
             Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
         };
 
-        let sink = manager
+        let state = match manager
             .streams
             .get(&bind.producer_id)
             .context(format!("Producer {:?} not found", bind.producer_id))?
+            .state
+            .lock()
+        {
+            Ok(guard) => guard,
+            Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
+        };
+
+        let sink = state
             .pipeline
             .inner_state_as_ref()
             .sinks
@@ -447,10 +506,18 @@ impl WebRTCSessionManagementInterface for Manager {
             Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
         };
 
-        let sink = manager
+        let state = match manager
             .streams
             .get(&bind.producer_id)
             .context(format!("Producer {:?} not found", bind.producer_id))?
+            .state
+            .lock()
+        {
+            Ok(guard) => guard,
+            Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
+        };
+
+        let sink = state
             .pipeline
             .inner_state_as_ref()
             .sinks
@@ -477,7 +544,11 @@ impl StreamManagementInterface<StreamStatus> for Manager {
             Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
         };
 
-        let stream_id = stream.id;
+        let stream_id = match stream.state.lock() {
+            Ok(guard) => guard.pipeline_id,
+            Err(error) => return Err(anyhow!("Failed locking a Mutex. Reason: {error}")),
+        };
+
         if manager.streams.insert(stream_id, stream).is_some() {
             return Err(anyhow!("Failed adding stream {stream_id:?}"));
         }
@@ -559,10 +630,20 @@ impl StreamManagementInterface<StreamStatus> for Manager {
         Ok(manager
             .streams
             .values()
-            .map(|stream| StreamStatus {
-                id: stream.id,
-                running: stream.pipeline.is_running(),
-                video_and_stream: stream.video_and_stream_information.clone(),
+            .filter_map(|stream| {
+                let state = match stream.state.lock() {
+                    Ok(guard) => guard,
+                    Err(error) => {
+                        error!("Failed locking a Mutex. Reason: {error}");
+                        return None;
+                    }
+                };
+
+                Some(StreamStatus {
+                    id: state.pipeline_id,
+                    running: state.pipeline.is_running(),
+                    video_and_stream: state.video_and_stream_information.clone(),
+                })
             })
             .collect())
     }

--- a/src/stream/manager.rs
+++ b/src/stream/manager.rs
@@ -5,10 +5,7 @@ use std::{
 
 use crate::{
     settings,
-    stream::{
-        gst::utils::wait_for_element_state, types::CaptureConfiguration,
-        webrtc::signalling_protocol::BindAnswer,
-    },
+    stream::{types::CaptureConfiguration, webrtc::signalling_protocol::BindAnswer},
     video::video_source,
 };
 use crate::{stream::sink::SinkInterface, video::types::VideoSourceType};
@@ -23,7 +20,6 @@ type ClonableResult<T> = Result<T, Arc<Error>>;
 
 use async_std::stream::StreamExt;
 use cached::proc_macro::cached;
-use gst::{prelude::*, traits::ElementExt};
 use tracing::*;
 
 use super::{
@@ -568,52 +564,12 @@ impl StreamManagementInterface<StreamStatus> for Manager {
             return Err(anyhow!("Already removed"));
         }
 
-        let mut stream = manager
+        manager
             .streams
             .remove(stream_id)
             .context(format!("Stream {stream_id:?} not found"))?;
         manager.update_settings();
         drop(manager);
-
-        let pipeline = &stream.pipeline.inner_state_as_ref().pipeline;
-        let pipeline_id = stream_id;
-        if let Err(error) = pipeline.post_message(gst::message::Eos::new()) {
-            error!("Failed posting Eos message into Pipeline' bus. Reason: {error:?}");
-        }
-
-        if let Err(error) = stream
-            .pipeline
-            .inner_state_as_ref()
-            .pipeline
-            .set_state(gst::State::Null)
-        {
-            error!("Failed setting Pipeline {pipeline_id:?} state to Null. Reason: {error:?}");
-        }
-        if let Err(error) = wait_for_element_state(
-            pipeline.upcast_ref::<gst::Element>(),
-            gst::State::Null,
-            100,
-            10,
-        ) {
-            let _ = pipeline.set_state(gst::State::Null);
-            error!("Failed setting Pipeline {pipeline_id:?} state to Null. Reason: {error:?}");
-        }
-
-        // Remove all Sinks
-        let sink_ids = &stream
-            .pipeline
-            .inner_state_as_ref()
-            .sinks
-            .keys()
-            .cloned()
-            .collect::<Vec<uuid::Uuid>>();
-        for sink_id in sink_ids {
-            if let Err(error) = stream.pipeline.remove_sink(sink_id) {
-                warn!(
-                    "Failed unlinking Sink {sink_id:?} from Pipeline {pipeline_id:?}. Reason: {error:?}"
-                );
-            }
-        }
 
         info!("Stream {stream_id} successfully removed!");
 

--- a/src/stream/mod.rs
+++ b/src/stream/mod.rs
@@ -6,8 +6,11 @@ pub mod sink;
 pub mod types;
 pub mod webrtc;
 
+use std::sync::{Arc, Mutex};
+
 use crate::mavlink::mavlink_camera::MavlinkCameraHandle;
 use crate::video::types::{VideoEncodeType, VideoSourceType};
+use crate::video::video_source::cameras_available;
 use crate::video_stream::types::VideoAndStreamInformation;
 
 use manager::Manager;
@@ -17,7 +20,7 @@ use types::*;
 use webrtc::signalling_protocol::PeerId;
 use webrtc::signalling_server::StreamManagementInterface;
 
-use anyhow::{anyhow, Result};
+use anyhow::{anyhow, Context, Result};
 
 use tracing::*;
 
@@ -25,7 +28,14 @@ use self::rtsp::rtsp_server::RTSP_SERVER_PORT;
 
 #[derive(Debug)]
 pub struct Stream {
-    pub id: PeerId,
+    state: Arc<Mutex<StreamState>>,
+    terminated: Arc<Mutex<bool>>,
+    _watcher_thread_handle: std::thread::JoinHandle<()>,
+}
+
+#[derive(Debug)]
+pub struct StreamState {
+    pub pipeline_id: PeerId,
     pub pipeline: Pipeline,
     pub video_and_stream_information: VideoAndStreamInformation,
     pub mavlink_camera: Option<MavlinkCameraHandle>,
@@ -34,6 +44,146 @@ pub struct Stream {
 impl Stream {
     #[instrument(level = "debug")]
     pub fn try_new(video_and_stream_information: &VideoAndStreamInformation) -> Result<Self> {
+        let pipeline_id = Manager::generate_uuid();
+
+        let state = Arc::new(Mutex::new(StreamState::try_new(
+            video_and_stream_information,
+            &pipeline_id,
+        )?));
+
+        let terminated = Arc::new(Mutex::new(false));
+        let terminated_cloned = terminated.clone();
+
+        let video_and_stream_information_cloned = video_and_stream_information.clone();
+        let state_cloned = state.clone();
+        let _watcher_thread_handle = std::thread::Builder::new()
+            .name(format!("Stream-{pipeline_id}"))
+            .spawn(move || {
+                Self::watcher(
+                    video_and_stream_information_cloned,
+                    pipeline_id,
+                    state_cloned,
+                    terminated_cloned,
+                )
+            })
+            .context(format!(
+                "Failed when spawing PipelineRunner thread for Pipeline {pipeline_id:#?}"
+            ))?;
+
+        Ok(Self {
+            state,
+            terminated,
+            _watcher_thread_handle,
+        })
+    }
+
+    #[instrument(level = "debug")]
+    fn watcher(
+        video_and_stream_information: VideoAndStreamInformation,
+        pipeline_id: uuid::Uuid,
+        state: Arc<Mutex<StreamState>>,
+        terminated: Arc<Mutex<bool>>,
+    ) {
+        // To reduce log size, each report we raise the report interval geometrically until a maximum value is reached:
+        let report_interval_mult = 2;
+        let report_interval_max = 60;
+        let mut report_interval = std::time::Duration::from_secs(1);
+        let mut last_report_time = std::time::Instant::now();
+
+        let mut video_and_stream_information = video_and_stream_information;
+
+        loop {
+            std::thread::sleep(std::time::Duration::from_millis(100));
+
+            if !state
+                .lock()
+                .unwrap()
+                .pipeline
+                .inner_state_as_ref()
+                .pipeline_runner
+                .is_running()
+            {
+                // If it's a camera, try to update the device
+                if let VideoSourceType::Local(_) = video_and_stream_information.video_source {
+                    let mut streams = vec![video_and_stream_information.clone()];
+                    let mut candidates = cameras_available();
+
+                    // Discards any source from other running streams, otherwise we'd be trying to create a stream from a device in use (which is not possible)
+                    let current_running_streams = manager::streams()
+                        .unwrap()
+                        .iter()
+                        .filter_map(|status| {
+                            status
+                                .running
+                                .then_some(status.video_and_stream.video_source.clone())
+                        })
+                        .collect::<Vec<VideoSourceType>>();
+                    candidates.retain(|candidate| !current_running_streams.contains(candidate));
+
+                    let should_report =
+                        std::time::Instant::now() - last_report_time >= report_interval;
+
+                    // Find the best candidate
+                    manager::update_devices(&mut streams, &mut candidates, should_report);
+                    video_and_stream_information = streams.first().unwrap().clone();
+
+                    // Check if the chosen video source is available
+                    match crate::video::video_source::get_video_source(
+                        video_and_stream_information
+                            .video_source
+                            .inner()
+                            .source_string(),
+                    ) {
+                        Ok(best_candidate) => {
+                            video_and_stream_information.video_source = best_candidate;
+                        }
+                        Err(error) => {
+                            if should_report {
+                                warn!("Failed to recreate the stream {pipeline_id:?}: {error:?}. Is the device connected? Trying again each second until the success or stream is removed. Next report in {report_interval:?} to reduce log size.");
+                                last_report_time = std::time::Instant::now();
+                                report_interval *= report_interval_mult;
+                                if report_interval
+                                    > std::time::Duration::from_secs(report_interval_max)
+                                {
+                                    report_interval =
+                                        std::time::Duration::from_secs(report_interval_max);
+                                }
+                            }
+
+                            std::thread::sleep(std::time::Duration::from_secs(1));
+                            continue;
+                        }
+                    }
+                }
+
+                // Try to recreate the stream
+                *state.lock().unwrap() = match StreamState::try_new(
+                    &video_and_stream_information,
+                    &pipeline_id,
+                ) {
+                    Ok(state) => state,
+                    Err(error) => {
+                        error!("Failed to recreate the stream {pipeline_id:?}: {error:#?}. Trying again in one second...");
+                        std::thread::sleep(std::time::Duration::from_secs(1));
+                        continue;
+                    }
+                };
+            }
+
+            if *terminated.lock().unwrap() {
+                debug!("Ending stream {pipeline_id:?}.");
+                break;
+            }
+        }
+    }
+}
+
+impl StreamState {
+    #[instrument(level = "debug")]
+    pub fn try_new(
+        video_and_stream_information: &VideoAndStreamInformation,
+        pipeline_id: &uuid::Uuid,
+    ) -> Result<Self> {
         if let Err(error) = validate_endpoints(video_and_stream_information) {
             return Err(anyhow!("Failed validating endpoints. Reason: {error:?}"));
         }

--- a/src/stream/mod.rs
+++ b/src/stream/mod.rs
@@ -54,7 +54,7 @@ impl Stream {
         };
 
         let mut stream = Stream {
-            id,
+            pipeline_id: *pipeline_id,
             pipeline,
             video_and_stream_information: video_and_stream_information.clone(),
             mavlink_camera,

--- a/src/stream/mod.rs
+++ b/src/stream/mod.rs
@@ -24,7 +24,11 @@ use anyhow::{anyhow, Context, Result};
 
 use tracing::*;
 
+use self::gst::utils::wait_for_element_state;
 use self::rtsp::rtsp_server::RTSP_SERVER_PORT;
+use self::sink::SinkInterface;
+
+use ::gst::prelude::*;
 
 #[derive(Debug)]
 pub struct Stream {
@@ -178,6 +182,12 @@ impl Stream {
     }
 }
 
+impl Drop for Stream {
+    fn drop(&mut self) {
+        *self.terminated.lock().unwrap() = true;
+    }
+}
+
 impl StreamState {
     #[instrument(level = "debug")]
     pub fn try_new(
@@ -188,8 +198,7 @@ impl StreamState {
             return Err(anyhow!("Failed validating endpoints. Reason: {error:?}"));
         }
 
-        let pipeline = Pipeline::try_new(video_and_stream_information)?;
-        let id = pipeline.inner_state_as_ref().pipeline_id;
+        let pipeline = Pipeline::try_new(video_and_stream_information, pipeline_id)?;
 
         // Only create the Mavlink Handle when mavlink is not disabled
         let mavlink_camera = match video_and_stream_information
@@ -203,7 +212,7 @@ impl StreamState {
             _ => MavlinkCameraHandle::try_new(video_and_stream_information).ok(),
         };
 
-        let mut stream = Stream {
+        let mut stream = StreamState {
             pipeline_id: *pipeline_id,
             pipeline,
             video_and_stream_information: video_and_stream_information.clone(),
@@ -270,6 +279,43 @@ impl StreamState {
         }
 
         Ok(stream)
+    }
+}
+
+impl Drop for StreamState {
+    fn drop(&mut self) {
+        let pipeline_state = self.pipeline.inner_state_as_ref();
+        let pipeline = &pipeline_state.pipeline;
+
+        if let Err(error) = pipeline.post_message(::gst::message::Eos::new()) {
+            error!("Failed posting Eos message into Pipeline bus. Reason: {error:?}");
+        }
+
+        if let Err(error) = pipeline.set_state(::gst::State::Null) {
+            error!("Failed setting Pipeline state to Null. Reason: {error:?}");
+        }
+        if let Err(error) = wait_for_element_state(
+            pipeline.upcast_ref::<::gst::Element>(),
+            ::gst::State::Null,
+            100,
+            10,
+        ) {
+            let _ = pipeline.set_state(::gst::State::Null);
+            error!("Failed setting Pipeline state to Null. Reason: {error:?}");
+        }
+
+        // Remove all Sinks
+        let pipeline_state = self.pipeline.inner_state_mut();
+        let sink_ids = &pipeline_state
+            .sinks
+            .keys()
+            .cloned()
+            .collect::<Vec<uuid::Uuid>>();
+        for sink_id in sink_ids {
+            if let Err(error) = pipeline_state.remove_sink(sink_id) {
+                warn!("Failed unlinking Sink {sink_id:?} from Pipeline. Reason: {error:?}");
+            }
+        }
     }
 }
 

--- a/src/stream/mod.rs
+++ b/src/stream/mod.rs
@@ -114,6 +114,11 @@ impl Stream {
             .pipeline_runner
             .start()?;
 
+        // Start all the sinks
+        for sink in stream.pipeline.inner_state_mut().sinks.values() {
+            sink.start()?
+        }
+
         Ok(stream)
     }
 }

--- a/src/stream/pipeline/fake_pipeline.rs
+++ b/src/stream/pipeline/fake_pipeline.rs
@@ -56,6 +56,9 @@ impl FakePipeline {
             }
         };
 
+        let filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}");
+        let sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}");
+
         // Fakes (videotestsrc) are only "video/x-raw" or "video/x-bayer",
         // and to be able to encode it, we need to define an available
         // format for both its src the next element's sink pad.
@@ -80,8 +83,8 @@ impl FakePipeline {
                     height = configuration.height,
                     interval_denominator = configuration.frame_interval.denominator,
                     interval_numerator = configuration.frame_interval.numerator,
-                    filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}"),
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    filter_name = filter_name,
+                    sink_tee_name = sink_tee_name,
                 )
             }
             VideoEncodeType::Yuyv => {
@@ -102,8 +105,8 @@ impl FakePipeline {
                     height = configuration.height,
                     interval_denominator = configuration.frame_interval.denominator,
                     interval_numerator = configuration.frame_interval.numerator,
-                    filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}"),
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    filter_name = filter_name,
+                    sink_tee_name = sink_tee_name,
                 )
             }
             VideoEncodeType::Mjpg => {
@@ -122,8 +125,8 @@ impl FakePipeline {
                     height = configuration.height,
                     interval_denominator = configuration.frame_interval.denominator,
                     interval_numerator = configuration.frame_interval.numerator,
-                    filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}"),
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    filter_name = filter_name,
+                    sink_tee_name = sink_tee_name,
                 )
             }
             unsupported => {

--- a/src/stream/pipeline/fake_pipeline.rs
+++ b/src/stream/pipeline/fake_pipeline.rs
@@ -25,7 +25,7 @@ pub struct FakePipeline {
 impl FakePipeline {
     #[instrument(level = "debug")]
     pub fn try_new(
-        pipeline_id: uuid::Uuid,
+        pipeline_id: &uuid::Uuid,
         video_and_stream_information: &VideoAndStreamInformation,
     ) -> Result<gst::Pipeline> {
         let configuration = match &video_and_stream_information

--- a/src/stream/pipeline/mod.rs
+++ b/src/stream/pipeline/mod.rs
@@ -238,6 +238,9 @@ impl PipelineState {
             format!("pipeline-{pipeline_id}-sink-{sink_id}-before-removing"),
         );
 
+        // Terminate the Sink
+        sink.eos();
+
         // Unlink the Sink
         sink.unlink(pipeline, pipeline_id)?;
 

--- a/src/stream/pipeline/mod.rs
+++ b/src/stream/pipeline/mod.rs
@@ -16,10 +16,8 @@ use gst::prelude::*;
 use crate::{
     stream::{
         gst::utils::wait_for_element_state,
-        manager::Manager,
         rtsp::rtsp_server::RTSPServer,
         sink::{Sink, SinkInterface},
-        webrtc::signalling_server::StreamManagementInterface,
     },
     video::types::VideoSourceType,
     video_stream::types::VideoAndStreamInformation,
@@ -98,7 +96,6 @@ pub struct PipelineState {
     pub sink_tee: gst::Element,
     pub sinks: HashMap<uuid::Uuid, Sink>,
     pub pipeline_runner: PipelineRunner,
-    _watcher_thread_handle: tokio::task::JoinHandle<()>,
 }
 
 pub const PIPELINE_SINK_TEE_NAME: &str = "SinkTee";
@@ -127,7 +124,6 @@ impl PipelineState {
             .context(format!("no element named {PIPELINE_SINK_TEE_NAME:#?}"))?;
 
         let pipeline_runner = PipelineRunner::try_new(&pipeline, pipeline_id, false)?;
-        let mut killswitch_receiver = pipeline_runner.get_receiver();
 
         pipeline.debug_to_dot_file_with_ts(
             gst::DebugGraphDetails::all(),
@@ -140,17 +136,6 @@ impl PipelineState {
             sink_tee,
             sinks: Default::default(),
             pipeline_runner,
-            _watcher_thread_handle: tokio::task::spawn(async move {
-                // Here we end the stream if any error is received. This should end all sessions too.
-                if let Ok(reason) = killswitch_receiver.recv().await {
-                    debug!("Killswitch received as {pipeline_id:#?} from PipelineState's watcher. Reason: {reason:#?}");
-                    if let Err(reason) = Manager::remove_stream(&pipeline_id) {
-                        warn!("Failed removing Pipeline {pipeline_id}. Reason: {reason}");
-                    } else {
-                        info!("Pipeline {pipeline_id} removed. Reason: {reason}");
-                    }
-                }
-            }),
         })
     }
 

--- a/src/stream/pipeline/mod.rs
+++ b/src/stream/pipeline/mod.rs
@@ -61,8 +61,11 @@ impl Pipeline {
     }
 
     #[instrument(level = "debug")]
-    pub fn try_new(video_and_stream_information: &VideoAndStreamInformation) -> Result<Self> {
-        let pipeline_state = PipelineState::try_new(video_and_stream_information)?;
+    pub fn try_new(
+        video_and_stream_information: &VideoAndStreamInformation,
+        pipeline_id: &uuid::Uuid,
+    ) -> Result<Self> {
+        let pipeline_state = PipelineState::try_new(video_and_stream_information, pipeline_id)?;
         Ok(match &video_and_stream_information.video_source {
             VideoSourceType::Gst(_) => Pipeline::Fake(FakePipeline {
                 state: pipeline_state,
@@ -103,9 +106,10 @@ pub const PIPELINE_FILTER_NAME: &str = "Filter";
 
 impl PipelineState {
     #[instrument(level = "debug")]
-    pub fn try_new(video_and_stream_information: &VideoAndStreamInformation) -> Result<Self> {
-        let pipeline_id = Manager::generate_uuid();
-
+    pub fn try_new(
+        video_and_stream_information: &VideoAndStreamInformation,
+        pipeline_id: &uuid::Uuid,
+    ) -> Result<Self> {
         let pipeline = match &video_and_stream_information.video_source {
             VideoSourceType::Gst(_) => {
                 FakePipeline::try_new(pipeline_id, video_and_stream_information)
@@ -131,7 +135,7 @@ impl PipelineState {
         );
 
         Ok(Self {
-            pipeline_id,
+            pipeline_id: *pipeline_id,
             pipeline,
             sink_tee,
             sinks: Default::default(),
@@ -163,13 +167,13 @@ impl PipelineState {
 
         // Link the Sink
         let pipeline = &self.pipeline;
-        sink.link(pipeline, &self.pipeline_id, tee_src_pad)?;
+        sink.link(pipeline, pipeline_id, tee_src_pad)?;
         let sink_id = &sink.get_id();
 
         // Start the pipeline if not playing yet
         if pipeline.current_state() != gst::State::Playing {
             if let Err(error) = pipeline.set_state(gst::State::Playing) {
-                sink.unlink(pipeline, &self.pipeline_id)?;
+                sink.unlink(pipeline, pipeline_id)?;
                 return Err(anyhow!(
                     "Failed starting Pipeline {pipeline_id}. Reason: {error:#?}"
                 ));
@@ -183,7 +187,7 @@ impl PipelineState {
             2,
         ) {
             let _ = pipeline.set_state(gst::State::Null);
-            sink.unlink(pipeline, &self.pipeline_id)?;
+            sink.unlink(pipeline, pipeline_id)?;
             return Err(anyhow!(
                 "Failed setting Pipeline {pipeline_id} to Playing state. Reason: {error:?}"
             ));

--- a/src/stream/pipeline/redirect_pipeline.rs
+++ b/src/stream/pipeline/redirect_pipeline.rs
@@ -19,7 +19,7 @@ pub struct RedirectPipeline {
 impl RedirectPipeline {
     #[instrument(level = "debug")]
     pub fn try_new(
-        pipeline_id: uuid::Uuid,
+        pipeline_id: &uuid::Uuid,
         video_and_stream_information: &VideoAndStreamInformation,
     ) -> Result<gst::Pipeline> {
         match &video_and_stream_information

--- a/src/stream/pipeline/redirect_pipeline.rs
+++ b/src/stream/pipeline/redirect_pipeline.rs
@@ -57,6 +57,8 @@ impl RedirectPipeline {
             .first()
             .context("Failed to access the fisrt endpoint")?;
 
+        let sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}");
+
         let description = match url.scheme() {
             "rtsp" => {
                 format!(
@@ -66,7 +68,7 @@ impl RedirectPipeline {
                         " ! tee name={sink_tee_name} allow-not-linked=true"
                     ),
                     location = url,
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    sink_tee_name = sink_tee_name,
                 )
             }
             "udp" => {
@@ -78,7 +80,7 @@ impl RedirectPipeline {
                     ),
                     address = url.host().context("UDP URL without host")?,
                     port = url.port().context("UDP URL without port")?,
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    sink_tee_name = sink_tee_name,
                 )
             }
             unsupported => {

--- a/src/stream/pipeline/runner.rs
+++ b/src/stream/pipeline/runner.rs
@@ -21,7 +21,7 @@ impl PipelineRunner {
     #[instrument(level = "debug")]
     pub fn try_new(
         pipeline: &gst::Pipeline,
-        pipeline_id: uuid::Uuid,
+        pipeline_id: &uuid::Uuid,
         allow_block: bool,
     ) -> Result<Self> {
         let pipeline_weak = pipeline.downgrade();
@@ -83,9 +83,7 @@ impl PipelineRunner {
     #[instrument(level = "debug")]
     fn runner(
         pipeline_weak: gst::glib::WeakRef<gst::Pipeline>,
-        pipeline_id: uuid::Uuid,
-        mut killswitch_receiver: broadcast::Receiver<String>,
-        mut start_signal_receiver: broadcast::Receiver<()>,
+        pipeline_id: &uuid::Uuid,
         allow_block: bool,
     ) -> Result<()> {
         let pipeline = pipeline_weak

--- a/src/stream/pipeline/runner.rs
+++ b/src/stream/pipeline/runner.rs
@@ -1,8 +1,9 @@
+use std::sync::{Arc, Mutex};
+
 use gst::prelude::*;
 
 use anyhow::{anyhow, Context, Result};
 
-use tokio::sync::broadcast;
 use tracing::*;
 
 use crate::stream::gst::utils::wait_for_element_state;
@@ -11,10 +12,8 @@ use crate::stream::gst::utils::wait_for_element_state;
 #[allow(dead_code)]
 pub struct PipelineRunner {
     pipeline_weak: gst::glib::WeakRef<gst::Pipeline>,
-    start_signal_sender: broadcast::Sender<()>,
-    killswitch_sender: broadcast::Sender<String>,
-    _killswitch_receiver: broadcast::Receiver<String>,
-    _watcher_thread_handle: std::thread::JoinHandle<()>,
+    start: Arc<Mutex<bool>>,
+    watcher_thread_handle: std::thread::JoinHandle<()>,
 }
 
 impl PipelineRunner {
@@ -25,37 +24,22 @@ impl PipelineRunner {
         allow_block: bool,
     ) -> Result<Self> {
         let pipeline_weak = pipeline.downgrade();
-        let (killswitch_sender, _killswitch_receiver) = broadcast::channel(1);
-        let watcher_killswitch_receiver = killswitch_sender.subscribe();
-        let (start_signal_sender, start_signal_receiver) = broadcast::channel(1);
+        let pipeline_id = *pipeline_id;
+
+        let start = Arc::new(Mutex::new(false));
 
         Ok(Self {
             pipeline_weak: pipeline_weak.clone(),
-            start_signal_sender,
-            killswitch_sender: killswitch_sender.clone(),
-            _killswitch_receiver,
-            _watcher_thread_handle: std::thread::Builder::new()
+            start: start.clone(),
+            watcher_thread_handle: std::thread::Builder::new()
                 .name(format!("PipelineRunner-{pipeline_id}"))
                 .spawn(move || {
-                    let mut reason = "Normal ending".to_string();
-                    if let Err(error) = PipelineRunner::runner(
-                        pipeline_weak,
-                        pipeline_id,
-                        watcher_killswitch_receiver,
-                        start_signal_receiver,
-                        allow_block,
-                    ) {
+                    if let Err(error) =
+                        PipelineRunner::runner(pipeline_weak, &pipeline_id, start, allow_block)
+                    {
                         error!("PipelineWatcher ended with error: {error}");
-                        reason = error.to_string();
                     } else {
-                        info!("PipelineWatcher ended with no error.");
-                    }
-
-                    // Any ending reason should interrupt the respective pipeline
-                    if let Err(reason) = killswitch_sender.send(reason) {
-                        error!("Failed to broadcast error from PipelineWatcher. Reason: {reason}");
-                    } else {
-                        info!("Error sent to killswitch channel!");
+                        info!("PipelineWatcher ended normally.");
                     }
                 })
                 .context(format!(
@@ -65,25 +49,21 @@ impl PipelineRunner {
     }
 
     #[instrument(level = "debug", skip(self))]
-    pub fn get_receiver(&self) -> broadcast::Receiver<String> {
-        self.killswitch_sender.subscribe()
-    }
-
-    #[instrument(level = "debug", skip(self))]
     pub fn start(&self) -> Result<()> {
-        self.start_signal_sender.send(())?;
+        *self.start.lock().unwrap() = true;
         Ok(())
     }
 
     #[instrument(level = "debug", skip(self))]
     pub fn is_running(&self) -> bool {
-        !self._watcher_thread_handle.is_finished()
+        !self.watcher_thread_handle.is_finished()
     }
 
     #[instrument(level = "debug")]
     fn runner(
         pipeline_weak: gst::glib::WeakRef<gst::Pipeline>,
         pipeline_id: &uuid::Uuid,
+        start: Arc<Mutex<bool>>,
         allow_block: bool,
     ) -> Result<()> {
         let pipeline = pipeline_weak
@@ -102,24 +82,11 @@ impl PipelineRunner {
         let mut lost_timestamps: usize = 0;
         let max_lost_timestamps: usize = 15;
 
-        let mut start_received = false;
-
         'outer: loop {
             std::thread::sleep(std::time::Duration::from_millis(100));
 
             // Wait the signal to start
-            if !start_received {
-                if let Err(error) = start_signal_receiver.try_recv() {
-                    match error {
-                        broadcast::error::TryRecvError::Empty => continue,
-                        _ => return Err(anyhow!("Failed receiving start signal: {error:?}")),
-                    }
-                }
-                debug!("Starting signal received in Pipeline {pipeline_id}");
-                start_received = true;
-            }
-
-            if pipeline.current_state() != gst::State::Playing {
+            if *start.lock().unwrap() && pipeline.current_state() != gst::State::Playing {
                 if let Err(error) = pipeline.set_state(gst::State::Playing) {
                     return Err(anyhow!(
                         "Failed setting Pipeline {pipeline_id} to Playing state. Reason: {error:?}"
@@ -216,11 +183,6 @@ impl PipelineRunner {
                         }
                         other_message => trace!("{other_message:#?}"),
                     };
-                }
-
-                if let Ok(reason) = killswitch_receiver.try_recv() {
-                    debug!("Killswitch received as {pipeline_id:#?} from PipelineRunner's watcher. Reason: {reason:#?}");
-                    break 'outer;
                 }
             }
         }

--- a/src/stream/pipeline/runner.rs
+++ b/src/stream/pipeline/runner.rs
@@ -98,10 +98,9 @@ impl PipelineRunner {
                     100,
                     5,
                 ) {
-                    error!(
+                    return Err(anyhow!(
                         "Failed setting Pipeline {pipeline_id} to Playing state. Reason: {error:?}"
-                    );
-                    continue;
+                    ));
                 }
             }
 

--- a/src/stream/pipeline/runner.rs
+++ b/src/stream/pipeline/runner.rs
@@ -141,7 +141,7 @@ impl PipelineRunner {
 
                     match msg.view() {
                         MessageView::Eos(eos) => {
-                            warn!("Received EndOfStream: {eos:#?}");
+                            debug!("Received EndOfStream: {eos:?}");
                             pipeline.debug_to_dot_file_with_ts(
                                 gst::DebugGraphDetails::all(),
                                 format!("pipeline-{pipeline_id}-eos"),

--- a/src/stream/pipeline/runner.rs
+++ b/src/stream/pipeline/runner.rs
@@ -115,16 +115,15 @@ impl PipelineRunner {
                                     && current_previous_position == position
                                 {
                                     lost_timestamps += 1;
-                                    warn!("Position did not change {lost_timestamps}");
-                                } else {
+                                } else if lost_timestamps > 0 {
                                     // We are back in track, erase lost timestamps
+                                    warn!("Position normalized, but didn't changed for {lost_timestamps} timestamps");
                                     lost_timestamps = 0;
                                 }
-
-                                if lost_timestamps > max_lost_timestamps {
-                                    warn!("Pipeline lost too many timestamps (max. was {max_lost_timestamps}).");
-                                    lost_timestamps = 0;
-                                    break 'inner;
+                                if lost_timestamps == 1 {
+                                    warn!("Position did not change for {lost_timestamps}, silently tracking until {max_lost_timestamps}, then the stream will be recreated");
+                                } else if lost_timestamps > max_lost_timestamps {
+                                    return Err(anyhow!("Pipeline lost too many timestamps (max. was {max_lost_timestamps})"));
                                 }
 
                                 Some(position)

--- a/src/stream/pipeline/runner.rs
+++ b/src/stream/pipeline/runner.rs
@@ -228,21 +228,3 @@ impl PipelineRunner {
         Ok(())
     }
 }
-
-impl Drop for PipelineRunner {
-    #[instrument(level = "debug", skip(self))]
-    fn drop(&mut self) {
-        if let Some(pipeline) = self.pipeline_weak.upgrade() {
-            pipeline.send_event(gst::event::Eos::new());
-        }
-
-        if let Err(reason) = self
-            .killswitch_sender
-            .send("PipelineRunner Dropped.".to_string())
-        {
-            warn!(
-                "Failed to send killswitch message while Dropping PipelineRunner. Reason: {reason:#?}"
-            );
-        }
-    }
-}

--- a/src/stream/pipeline/v4l_pipeline.rs
+++ b/src/stream/pipeline/v4l_pipeline.rs
@@ -22,7 +22,7 @@ pub struct V4lPipeline {
 impl V4lPipeline {
     #[instrument(level = "debug")]
     pub fn try_new(
-        pipeline_id: uuid::Uuid,
+        pipeline_id: &uuid::Uuid,
         video_and_stream_information: &VideoAndStreamInformation,
     ) -> Result<gst::Pipeline> {
         let configuration = match &video_and_stream_information

--- a/src/stream/pipeline/v4l_pipeline.rs
+++ b/src/stream/pipeline/v4l_pipeline.rs
@@ -47,6 +47,8 @@ impl V4lPipeline {
         let height = configuration.height;
         let interval_numerator = configuration.frame_interval.numerator;
         let interval_denominator = configuration.frame_interval.denominator;
+        let filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}");
+        let sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}");
 
         let description = match &configuration.encode {
             VideoEncodeType::H264 => {
@@ -63,8 +65,8 @@ impl V4lPipeline {
                     height = height,
                     interval_denominator = interval_denominator,
                     interval_numerator = interval_numerator,
-                    filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}"),
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    filter_name = filter_name,
+                    sink_tee_name = sink_tee_name,
                 )
             }
             VideoEncodeType::Yuyv => {
@@ -81,8 +83,8 @@ impl V4lPipeline {
                     height = height,
                     interval_denominator = interval_denominator,
                     interval_numerator = interval_numerator,
-                    filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}"),
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    filter_name = filter_name,
+                    sink_tee_name = sink_tee_name
                 )
             }
             VideoEncodeType::Mjpg => {
@@ -99,8 +101,8 @@ impl V4lPipeline {
                     height = height,
                     interval_denominator = interval_denominator,
                     interval_numerator = interval_numerator,
-                    filter_name = format!("{PIPELINE_FILTER_NAME}-{pipeline_id}"),
-                    sink_tee_name = format!("{PIPELINE_SINK_TEE_NAME}-{pipeline_id}"),
+                    filter_name = filter_name,
+                    sink_tee_name = sink_tee_name
                 )
             }
             unsupported => {

--- a/src/stream/sink/image_sink.rs
+++ b/src/stream/sink/image_sink.rs
@@ -69,7 +69,7 @@ pub struct ImageSink {
     tee_src_pad: Option<gst::Pad>,
     flat_samples_sender: tokio::sync::broadcast::Sender<ClonableResult<FlatSamples<Vec<u8>>>>,
     pad_blocker: Arc<Mutex<Option<gst::PadProbeId>>>,
-    _pipeline_runner: PipelineRunner,
+    pipeline_runner: PipelineRunner,
     thumbnails: Arc<Mutex<CachedThumbnails>>,
 }
 impl SinkInterface for ImageSink {
@@ -272,6 +272,18 @@ impl SinkInterface for ImageSink {
         Err(anyhow!(
             "Not available. Reason: Image Sink doesn't provide endpoints"
         ))
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn start(&self) -> Result<()> {
+        self.pipeline_runner.start()
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn eos(&self) {
+        if let Err(error) = self.pipeline.post_message(gst::message::Eos::new()) {
+            error!("Failed posting Eos message into Sink bus. Reason: {error:?}");
+        }
     }
 }
 
@@ -491,7 +503,7 @@ impl ImageSink {
             return Err(anyhow!("Failed linking ImageSink's elements: {link_err:?}"));
         }
 
-        let _pipeline_runner = PipelineRunner::try_new(&pipeline, sink_id, true)?;
+        let pipeline_runner = PipelineRunner::try_new(&pipeline, &sink_id, true)?;
 
         // Start the pipeline in Pause, because we want to wait the snapshot
         if let Err(state_err) = pipeline.set_state(gst::State::Paused) {
@@ -511,7 +523,7 @@ impl ImageSink {
             tee_src_pad: Default::default(),
             flat_samples_sender,
             pad_blocker,
-            _pipeline_runner,
+            pipeline_runner,
             thumbnails: Default::default(),
         })
     }

--- a/src/stream/sink/mod.rs
+++ b/src/stream/sink/mod.rs
@@ -38,6 +38,12 @@ pub trait SinkInterface {
     ///
     /// For a better grasp of SDP parameters, read [here](https://www.iana.org/assignments/sdp-parameters/sdp-parameters.xhtml)
     fn get_sdp(&self) -> Result<gst_sdp::SDPMessage>;
+
+    /// Start the Sink
+    fn start(&self) -> Result<()>;
+
+    /// Terminates the Sink
+    fn eos(&self);
 }
 
 #[enum_dispatch(SinkInterface)]

--- a/src/stream/sink/rtsp_sink.rs
+++ b/src/stream/sink/rtsp_sink.rs
@@ -209,6 +209,14 @@ impl SinkInterface for RtspSink {
             "Not available. Reason: RTSP Sink should only be connected from its RTSP endpoint."
         ))
     }
+
+    #[instrument(level = "debug", skip(self))]
+    fn start(&self) -> Result<()> {
+        Ok(())
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn eos(&self) {}
 }
 
 impl RtspSink {

--- a/src/stream/sink/udp_sink.rs
+++ b/src/stream/sink/udp_sink.rs
@@ -18,7 +18,7 @@ pub struct UdpSink {
     udpsink_sink_pad: gst::Pad,
     tee_src_pad: Option<gst::Pad>,
     addresses: Vec<url::Url>,
-    _pipeline_runner: PipelineRunner,
+    pipeline_runner: PipelineRunner,
 }
 impl SinkInterface for UdpSink {
     #[instrument(level = "debug", skip(self))]
@@ -253,6 +253,18 @@ impl SinkInterface for UdpSink {
 
         Ok(sdp)
     }
+
+    #[instrument(level = "debug", skip(self))]
+    fn start(&self) -> Result<()> {
+        self.pipeline_runner.start()
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn eos(&self) {
+        if let Err(error) = self.pipeline.post_message(gst::message::Eos::new()) {
+            error!("Failed posting Eos message into Sink bus. Reason: {error:?}");
+        }
+    }
 }
 
 impl UdpSink {
@@ -335,7 +347,7 @@ impl UdpSink {
             return Err(anyhow!("Failed linking UdpSink's elements: {link_err:?}"));
         }
 
-        let _pipeline_runner = PipelineRunner::try_new(&pipeline, sink_id, false)?;
+        let pipeline_runner = PipelineRunner::try_new(&pipeline, &sink_id, false)?;
 
         // Start the pipeline
         if let Err(state_err) = pipeline.set_state(gst::State::Playing) {
@@ -354,7 +366,7 @@ impl UdpSink {
             udpsink_sink_pad,
             addresses,
             tee_src_pad: Default::default(),
-            _pipeline_runner,
+            pipeline_runner,
         })
     }
 }

--- a/src/stream/sink/webrtc_sink.rs
+++ b/src/stream/sink/webrtc_sink.rs
@@ -113,6 +113,7 @@ impl SinkInterface for WebRTCSink {
         self.0.lock().unwrap().get_id()
     }
 
+    #[instrument(level = "debug", skip(self))]
     fn get_sdp(&self) -> Result<gst_sdp::SDPMessage> {
         self.0.lock().unwrap().get_sdp()
     }

--- a/src/stream/sink/webrtc_sink.rs
+++ b/src/stream/sink/webrtc_sink.rs
@@ -116,6 +116,16 @@ impl SinkInterface for WebRTCSink {
     fn get_sdp(&self) -> Result<gst_sdp::SDPMessage> {
         self.0.lock().unwrap().get_sdp()
     }
+
+    #[instrument(level = "debug", skip(self))]
+    fn start(&self) -> Result<()> {
+        self.0.lock().unwrap().start()
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn eos(&self) {
+        self.0.lock().unwrap().eos()
+    }
 }
 
 impl WebRTCSink {
@@ -659,6 +669,18 @@ impl SinkInterface for WebRTCSinkInner {
         Err(anyhow!(
             "Not available: WebRTC Sink should only be connected by means of its Signalling protocol."
         ))
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn start(&self) -> Result<()> {
+        Ok(())
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn eos(&self) {
+        if let Err(error) = self.webrtcbin.post_message(gst::message::Eos::new()) {
+            error!("Failed posting Eos message into Sink bus. Reason: {error:?}");
+        }
     }
 }
 

